### PR TITLE
MRG, FIX: allow arbitrary offset for {Epochs,Evoked}.shift_time

### PR DESCRIPTION
**Pull request by drammock**
- allow `Epochs.shift_time` and `Evoked.shift_time` to shift time by amounts that are not integer-multiples of the sampling period
- refactor so the same code is used for both classes

cc @jjnurminen @larsoner @agramfort

### CODE REVIEW

**Comment by larsoner** [comment]
Something bad has happened in `test_stim.py`:

https://travis-ci.org/mne-tools/mne-python/jobs/605244903

**Comment by drammock** [comment]
yep. I was noticing that for some reason `round` is returning floats, even though it's supposed to return `int`s when you don't specify `ndigits`.  I'll push a fix.

File: mne/evoked.py [hunk 1]
```diff
@@ -278,40 +277,7 @@ def decimate(self, decim, offset=0):
...
-        times = self.times
-        sfreq = self.info['sfreq']
-
-        offset = self.first if relative else 0
-
-        self.first = int(tshift * sfreq) + offset
-        self.last = self.first + len(times) - 1
-        self.times = np.arange(self.first, self.last + 1,
-                               dtype=np.float) / sfreq
+        # self._update_first_last()
```
Comment (drammock) [inline mne/evoked.py]: I would like to replace the two preceding lines in `decimate` with `_update_first_last()` but it breaks `mne/tests/test_evoked.py:test_decim()`, by changing the sample numbers.  Currently in the test, an Evoked w/ 20 timepoints (first=-10, last=9, times from -1 to 0.9) is decimated by a factor of 6, yielding first=-1, last=2, times=[-1, -0.4, 0.2, 0.8]. This means that "sample zero" was the sample associated with time=-0.4, which is *farther away from actual time zero* than the subsequent sample (the one associated with time 0.2).  So unless there is a convention that "sample zero" is always associated with the negative time in cases where no sample is exactly at time zero, I would argue that the test as it currently stands is wrong: if decimation yields times=[-1, -0.4, 0.2, 0.8], then "sample zero" should be the sample at time 0.2.  @larsoner WDYT?

Comment (agramfort) [reply mne/evoked.py]: I would say yes that sample zero should be the one closest to 0.

File: mne/tests/test_evoked.py [hunk 2]
```diff
@@ -252,6 +261,23 @@ def test_shift_time_evoked():
...
+
+    # test handling of Evoked.first, Evoked.last
+    ave = read_evokeds(fname, 0)
+    first_last = np.array([ave.first, ave.last])
+    # should shift by 0 samples
+    ave.shift_time(1e-6)
+    assert_array_equal(first_last, np.array([ave.first, ave.last]))
+    # should shift by 57 samples
+    ave.shift_time(57. / ave.info['sfreq'])
+    assert_array_equal(first_last + 57, np.array([ave.first, ave.last]))
```
Comment (larsoner) [inline mne/tests/test_evoked.py]: I suspect that something bad is going to happen on I/O round-trip following a shift, specifically for `Evoked` files. We always write `FIFF_FIRST_SAMPLE` and `FIFF_LAST_SAMPLE`, but never `FIFF_FIRST_TIME` and `FIFF_NO_SAMPLES`:
```
            # First and last sample
            write_int(fid, FIFF.FIFF_FIRST_SAMPLE, e.first)
            write_int(fid, FIFF.FIFF_LAST_SAMPLE, e.last)
```
Then on read our decision about times goes:
```
        if first is not None:
            nsamp = last - first + 1
        elif first_time is not None:
            first = int(round(first_time * info['sfreq']))
            last = first + nsamp
    ...
    times = np.arange(first, last + 1, dtype=np.float) / info['sfreq']
```

In the spirit of TDD, can you confirm this by adding a test I/O round-tripping your `1e-6`-shifted `evoked` instance will remove the shift?

To fix this, we can instead:

1. Always also write `FIFF_FIRST_TIME` and `FIFF_NO_SAMPLES` (in addition to / keeping `FIFF_FIRST_SAMPLE` and `FIFF_LAST_SAMPLE`
2. Prefer to use `FIFF_FIRST_TIME` and `FIFF_NO_SAMPLES`, if available, and if not available, fall back to `FIFF_FIRST_SAMPLE` and `FIFF_LAST_SAMPLE`

This should be backward compatible but also work for your sub-sample shift case.

Comment (drammock) [reply mne/tests/test_evoked.py]: @larsoner see if you're satisfied with the new read logic and the updated tests.  Times are stable through save/load roundtrip up to 8 decimal places (in my ad hoc testing sometimes 9, never 10).  Seems good enough to me but I trust your judgment here.

Comment (larsoner) [reply mne/tests/test_evoked.py]: Yes looks good. But I realized there is a tiny simplification / DRY we can get from no longer passing `first/last` out of `_read_evoked`, pushing a commit now. See if you think it's good then we should be ready for @agramfort to look and merge.

I also pushed a commit to fix a failing test by specifying `atol`.

**Comment by drammock** [comment]
thanks @larsoner for catching those things.  I had noticed early on about `first` and `last` in the dipole code, but forgot to go back to it.  CIs are happy and your changes LGTM.  +1 for merge.

**Comment by agramfort** [comment]
what's new page needs update. thx

**Comment by larsoner** [comment]
Thanks @drammock

**Comment by drammock** [comment]
LOL I was just pushing the "squash and merged" button

**Pull request merged**